language/sygus.tex

```diff
@@ -1068,29 +1084,62 @@ \subsection{Asserting Synthesis Constraints and Assumptions}
 where $(x_1, \ldots, x_n)$ and $(y_1, \ldots, y_n)$
 are tuples of variables of sort $(\sigma_1, \ldots, \sigma_n)$ and
 $\varphi_{pre}$, $\varphi_{trans}$ and $\varphi_{post}$ are formulas.
-Given that this command is well-formed, given the above definitions,
+
+When this command is well-formed, given the above definitions,
 this command is syntax sugar for:
-\[
-\begin{array}{l}
-   \paren{\vardeclkwd\mbox{ }v_1\mbox{ }\sigma_1}\\
-   \paren{\vardeclkwd\mbox{ }v'_1\mbox{ }\sigma_1}\\
-   \ldots\\
-   \paren{\vardeclkwd\mbox{ }v_n\mbox{ }\sigma_n}\\
-   \paren{\vardeclkwd\mbox{ }v'_n\mbox{ }\sigma_n}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S_{pre}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{\texttt{and}\mbox{ }\paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{trans}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}\mbox{ }
-   \paren{S\mbox{ }v'_1\mbox{ }\ldots\mbox{ }v'_n}}}\\
-   \paren{\constraintkwd\mbox{ }\paren{\texttt{=>}\mbox{ }
-   \paren{S\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}\mbox{ }
-   \paren{S_{post}\mbox{ }v_1\mbox{ }\ldots\mbox{ }v_n}}}\\
-\end{array}
-\]
+\begin{lstlisting}[language=SyGuS, numbers=none, basicstyle=\ttfamily\small]
+($\vardeclkwd$ $v_1$ $\sigma_1$)
+($\vardeclkwd$ $v'_1$ $\sigma_1$)
+$\ldots$
+($\vardeclkwd$ $v_n$ $\sigma_n$)
+($\vardeclkwd$ $v'_n$ $\sigma_n$)
+($\constraintkwd$ (=> ($S_{pre}$ $v_1$ $\ldots$ $v_n$) ($S$ $v_1$ $\ldots$ $v_n$)))
+($\constraintkwd$ (=> (and ($S$ $v_1$ $\ldots$ $v_n$) ($S_{trans}$ $v_1$ $\ldots$ $v_n$ $v'_1$ $\ldots$ $v'_n$)) ($S$ $v'_1$ $\ldots$ $v'_n$)))
+($\constraintkwd$ (=> ($S$ $v_1$ $\ldots$ $v_n$) ($S_{post}$ $v_1$ $\ldots$ $v_n$)))
+\end{lstlisting}
 where $v_1, v'_1, \ldots, v_n, v'_n$ are fresh symbols.
 
+\item \sygusinline|($\constraintchckwd$ (($x_1$ $\sigma_1$) $\ldots$ ($x_m$ $\sigma_m$)) $t_1$ $\ldots$ $t_n$ $t_{head}$)|
+
+This command adds a constraint to the current state,
+that corresponds to the constrained Horn clause
+$\forall x_1 : \sigma_1, \ldots, x_m : \sigma_m \ldotp t_{body} \implies t_{head}$,
+where $$
+t_{body} \triangleq
+\begin{cases}
+  (t_1 \land \ldots \land t_n)  & n > 0\\
+  \truekwd                      & n = 0
+\end{cases}
+$$
+This is a generalization of the $\constraintinvkwd$ command,
```

Comment:
> CHC variables can only be bound to the predicate within a single CHC constraint? Is that right?

Yes, this is right.

> If they are only bound to a single CHC constraint, what happens if I write this:
> ```
> (declare-var x Int)
> (constraint some predicate that uses x)
> (chc-constraint ((x Int)) tbody => thead)
> ```

If we desugar it for the SyGuS-general mode, then we would generate:
```
(declare-var x Int)
(some constraint on x)

(declare-var v_x Int)
(define-fun Body ((x Int)) tbody)
(define-fun Head ((x Int)) thead)
(constraint (=> (Body v_x) (Head v_x)))
```
You are correct about two variables: a fresh variable `v_x` local to the CHC, and `x` used in the non-CHC constraint.

However, for SyGuS input file to be valid in the CHC-solving mode, it must not use any other constraint-introducing statements besides `chc-constraint`.

Reply:
Ok, then I think this sentence needs changing:

> This command adds a constraint to the current state, that corresponds to the constrained Horn clause ∀x1 : σ1,...,xm : σm.tbody =⇒ thead

I think the correct thing to say is :

> This command adds a constraint  `tbody =⇒ thead`  to the current state and variables `x1,.., xm` to the list of universal variables. 

What you've written currently says:
- ` ∀x1 : σ1,...,xm : σm.tbody =⇒ thead` is added to the current list of constraints. 
- Then the synthesis conjecture, defined in section 3 is:
'`exists f, forall v. (alpha) => (rho)`', where '`rho`' is the list fo constraints, 
- so you've added `∀x1 : σ1,...,xm : σm.tbody =⇒ thead` to `rho`. 

Then, the synthesis conjecture you would end up with for the horn clauses is:
'`exists f, forall v. (alpha) => (∀x1 : σ1,...,xm : σm.tbody =⇒ thead)`',

Reply:
I see. Thank you! That makes sense. I will update this.

Reply:
Updated this. Please see if it looks okay now.

Reply:
This looks correct now. 

You might want to say that t_1.. t_n and t_head must be well-sorted terms of sort Bool. 

I also wonder why you need to define F_body and F_head? Why not just make
~~~
(chc-constraint ((x1 σ1) . . . (xm σm)) t1 . . . tn t_head)
~~~
syntactic sugar for this?
~~~
(declare-var x1 σ1 )
...
(declare-var xm σm )
(constraint (=> (t_body) (t_head)))
~~~

Is there something subtley different about that?

Reply:
It's because of possible name conflicts. Consider the following valid `chc-constraint`s:

```
(chc-constraint ((x σ)) t_body1 t_head1)
(chc-constraint ((x σ)) t_body2 t_head2)
```

If we don't introduce fresh variables, then both would end up doing `(declare-var x σ)`.

Reply:
ah ok